pkg/dependencies/dependents.go

```diff
@@ -0,0 +1,230 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dependencies
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	model "github.com/guacsec/guac/pkg/assembler/clients/generated"
+	"github.com/guacsec/guac/pkg/handler/collector/deps_dev"
+	"github.com/guacsec/guac/pkg/misc/depversion"
+
+	"github.com/Khan/genqlient/graphql"
+)
+
+type dependencyNode struct {
+	dependents map[string]bool // map of the Name of the dependent to whether it is a dependent of the current node
+}
+
+type PackageName struct {
+	Name           string
+	DependentCount int
+}
+
+// GetDependenciesBySortedDependentCnt retrieves all dependents for each package and returns a sorted list of
+// PackageName, where each PackageName contains the name of the package and the number of its dependents.
+// The list is sorted in descending order based on the DependentCount, so packages with the most dependents come first.
+// This function leverages the findDependentsOfDependencies function to construct a map of all packages and their dependents,
+// then processes this map to create a slice of PackageName, which is then sorted.
+//
+// Returns:
+// - A slice of PackageName, sorted by DependentCount in descending order.
+// - An error
+func GetDependenciesBySortedDependentCnt(ctx context.Context, gqlClient graphql.Client) ([]PackageName, error) {
+	packages, err := findDependentsOfDependencies(ctx, gqlClient)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get dependents: %v", err)
+	}
+
+	var packagesArr []PackageName
+
+	for n, d := range packages {
+		packagesArr = append(packagesArr, PackageName{Name: n, DependentCount: len(d.dependents)})
+	}
+
+	sort.Slice(packagesArr, func(i, j int) bool {
+		return packagesArr[i].DependentCount > packagesArr[j].DependentCount
+	})
+	return packagesArr, nil
+}
+
+// findDependentsOfDependencies queries the GraphQL endpoint to retrieve all SBOMs and constructs
+// a map of dependencyNode, where each node represents a package and its dependents. This function is designed to
+// identify and map out the relationships between packages based on the SBOM data. It filters out inconsistent data
+// from "deps.dev" and handles packages with and without specific version IDs. The function employs a breadth-first search
+// (BFS) algorithm to traverse the dependency graph and populate the map with packages and their respective dependents.
+//
+// Returns:
+//   - A map where the key is a string representing the package name, and the value is a dependencyNode containing
+//     the dependents of the package.
+//   - An error
+func findDependentsOfDependencies(ctx context.Context, gqlClient graphql.Client) (map[string]dependencyNode, error) {
```

Comment:
can you help break out this function and refactor some of the blocks out to aid readability?

Reply:
I refactored out part of the function into `updatePackagesAndNames`.